kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -715,6 +737,48 @@ impl SocketEventObserver for StreamSocket {
     }
 }
 
+impl State {
+    /// Calls `f` on raw socket.
+    ///
+    /// Note that for listening socket, this `f` is called on all backlog sockets in `Listen` State.
+    /// That is to say, f won't be called on backlog sockets in `SynReceived` or `Established` state.
+    fn for_raw_socket(&mut self, f: impl Fn(&mut RawTcpSocket)) {
+        match self {
+            State::Init(init_stream) => init_stream.for_raw_socket(f),
+            State::Connecting(connecting_stream) => connecting_stream.for_raw_socket(f),
+            State::Connected(connected_stream) => connected_stream.for_raw_socket(f),
+            State::Listen(listen_stream) => listen_stream.for_each_listen_backlog(f),
+        }
+    }
+}
+
+impl SetSocketLevelOption for State {
+    fn set_keep_alive(&mut self, keep_alive: bool) {
+        /// The keepalive interval.
+        ///
+        /// The linux value can be found at `/proc/sys/net/ipv4/tcp_keepalive_intvl`,
+        /// which is by default 75 seconds for most Linux distributions.
+        const KEEPALIVE_INTERVAL: aster_bigtcp::time::Duration =
+            aster_bigtcp::time::Duration::from_secs(75);
+
+        let set_keep_alive = |raw_socket: &mut RawTcpSocket| {
+            if keep_alive {
+                raw_socket.set_keep_alive(Some(KEEPALIVE_INTERVAL))
+            } else {
+                raw_socket.set_keep_alive(None);
+            }
+        };
+        match self {
+            State::Init(init_stream) => init_stream.for_raw_socket(set_keep_alive),
+            State::Connecting(connecting_stream) => {
+                connecting_stream.for_raw_socket(set_keep_alive)
+            }
+            State::Connected(connected_stream) => connected_stream.for_raw_socket(set_keep_alive),
+            State::Listen(listen_stream) => listen_stream.for_each_backlog(set_keep_alive),
```

Comment:
I noticed the code earlier. This is because setting the keep-alive option for a listening socket is useless on Linux. 

On Linux, the keep-alive flag is actually set when a new connection request arrives, leading to the creation of a new backlog socket to handle the request.

Refer to the following link for details: https://elixir.bootlin.com/linux/v6.0.9/source/net/ipv4/tcp_minisocks.c#L453

The new socket inherits the socket flags from the listening socket, including the keep-alive flag. Since our backlog sockets are created in advance, we need to set this flag for each backlog socket. 

However, the original implementation is incorrect. I discovered that disabling keep-alive on a listening socket actually does not take effect (I couldn't find any Linux source code confirming this behavior, but a new test case corroborates it). Any further accepted socket will also have keep-alive enabled. Therefore, the keep-alive on `ListenStream` should, like other methods, only affect backlog sockets in the Listen state.

Reply:
Thank you for pointing that out! I've fixed the issue, and I've discovered that keep-alive behaves just like TCP_NODELAY; only the backlog socket in the listen state is affected. This is consistent with the Linux source code.